### ISSUE TITLE
cStor SPC upgrade job fails because container does not start quickly enough

### ISSUE THREAD

**Issue by sgielen** [openebs/openebs#3062]
## Description
I am running an upgrade job on an arm64 cluster. Because the nodes are slow, it typically takes about 10 minutes for a cStor disk pool pod to restart with a new image. Most of this is in image pull, and then some in container creation. On amd64, this probably always happens quickly enough, but it seems to time out even though the process is still ongoing.

## Expected Behavior
I would expect the upgrade job to look at Pods in the pool Deployment and look at their current status. E.g., if the Pod is still in ContainerCreating, it has not yet failed to come up.

## Current Behavior
Currently, the upgrade job fails:

```
I0614 11:12:18.144521       1 cstor_spc.go:100] Started upgrading storagePoolClaim{cstor-disk-pool} from 1.9.0 to 1.11.0
I0614 11:12:18.669371       1 csp_upgrade.go:193] csp deployment cstor-disk-pool-una0 already in 1.11.0 version
I0614 11:12:18.695169       1 csp_upgrade.go:153] patched csp cstor-disk-pool-una0
I0614 11:12:18.695330       1 csp_upgrade.go:277] Verifying the reconciliation of version for cstor-disk-pool-una0 { phase:Healthy }
I0614 11:12:28.759719       1 csp_upgrade.go:380] Upgrade Successful for csp cstor-disk-pool-una0
I0614 11:12:29.295012       1 csp_upgrade.go:193] csp deployment cstor-disk-pool-7it6 already in 1.11.0 version
I0614 11:12:29.328885       1 csp_upgrade.go:153] patched csp cstor-disk-pool-7it6
I0614 11:12:29.328973       1 csp_upgrade.go:277] Verifying the reconciliation of version for cstor-disk-pool-7it6 { phase:Healthy }
I0614 11:12:39.352632       1 csp_upgrade.go:277] Verifying the reconciliation of version for cstor-disk-pool-7it6 { phase:Healthy }
I0614 11:12:49.422336       1 csp_upgrade.go:277] Verifying the reconciliation of version for cstor-disk-pool-7it6 { phase:Healthy }
I0614 11:12:59.565449       1 csp_upgrade.go:380] Upgrade Successful for csp cstor-disk-pool-7it6
I0614 11:13:00.149221       1 helper.go:69] Patching deployment cstor-disk-pool-eiq0
E0614 11:19:55.820252       1 cstor_spc.go:114] failed to patch deployment cstor-disk-pool-eiq0: failed to rollout because replica update in-progress: 0 of 1 updated replicas are available
F0614 11:19:55.820395       1 cstor_spc.go:78] Failed to upgrade cStor SPC cstor-disk-pool:
```

Even though the pool is still starting:

```
$ kubectl get pods -n openebs | grep cstor-disk-pool
cstor-disk-pool-una0-797496786c-xrtvk                             3/3     Running             0          112m
cstor-disk-pool-eiq0-6fdc6d9958-48qq2                             0/3     ContainerCreating   0          6m25s
cstor-disk-pool-7it6-5b5447466f-5llc8                             3/3     Running             0          118m
```

The Job automatically restarts, but exits again:

```
$ kubectl logs -n openebs -f cstor-spc-1901110-20200614-t8ht4
I0614 11:20:05.280791       1 cstor_spc.go:100] Started upgrading storagePoolClaim{cstor-disk-pool} from 1.9.0 to 1.11.0
E0614 11:20:06.281915       1 cstor_spc.go:114] failed to get deployment for csp cstor-disk-pool-eiq0: cstor-disk-pool-eiq0 deployment pods are not in running state expected: 1 got: 0
F0614 11:20:06.282135       1 cstor_spc.go:78] Failed to upgrade cStor SPC cstor-disk-pool:
```

Eventually, I have to delete the Job and recreate it when the Pod is Running again.

## Possible Solution
I would suggest the upgrader looks at the state of the Pod it's waiting for. If it's ContainerCreating, perhaps even checking the `describe` / Events to see in what phase of ContainerCreating it is, it keeps waiting. 

## Steps to Reproduce
- Have a cStor pool running on an older version
- Ensure cStor pool pods come up slowly. E.g., host images on a very slow registry, patch Kubelet to stay in ContainerCreating for a long time, or run on slower arm64 nodes.
- Try to perform an upgrade.

## Your Environment
* `kubectl get nodes`:
```
NAME       STATUS   ROLES    AGE    VERSION
kathleen   Ready    master   228d   v1.17.4+k3s1
karen      Ready    worker   225d   v1.17.4+k3s1
josje      Ready    <none>   33d    v1.17.4+k3s1
kristel    Ready    <none>   45d    v1.17.4+k3s1
```
* OpenEBS pods (version 1.9.0):
```
$ kubectl get pods -n openebs -l openebs.io/version=1.9.0 -o wide
NAME                                                              READY   STATUS        RESTARTS   AGE   IP            NODE       NOMINATED NODE   READINESS GATES
pvc-318b2f41-40db-4ee7-8564-4c8677665bbf-target-55cb656944cgrz5   3/3     Running       0          23d   10.42.3.60    josje      <none>           <none>
pvc-20bffc04-703d-44b0-8820-e124a9c0bc51-target-566b7f5f7fsff9b   3/3     Running       0          22d   10.42.3.65    josje      <none>           <none>
openebs-snapshot-operator-84fb87cf4c-q4f4v                        1/2     Terminating   46         45d   10.42.0.208   kathleen   <none>           <none>
pvc-1c329d55-f024-4306-8489-0efbd65f58c5-target-69bdf7d56-wp75s   3/3     Running       0          42d   10.42.0.228   kathleen   <none>           <none>
pvc-614627ce-bcf9-4d9e-941a-9408769679d7-target-5d4fb856b4x96qc   3/3     Running       0          42d   10.42.0.226   kathleen   <none>           <none>
pvc-f39179d6-2121-439d-ae18-bd4bd8f47f8d-target-78f9b46998xqt99   3/3     Running       0          42d   10.42.0.227   kathleen   <none>           <none>
pvc-b72ec2de-e301-4667-92cc-43d089dd2e02-target-5f6447cdf5lk2zc   3/3     Running       0          42d   10.42.0.230   kathleen   <none>           <none>
pvc-df6179f9-d398-442e-b3fb-aa1f859d43f7-target-875dc5d79-7q8wg   3/3     Running       1          42d   10.42.0.229   kathleen   <none>           <none>
```
  * The Terminating pod is due to https://github.com/rancher/k3s/issues/1901
* OpenEBS pods (version 1.11.0):
```
$ kubectl get pods -n openebs -l openebs.io/version=1.11.0 -o wide
NAME                                          READY   STATUS             RESTARTS   AGE     IP                NODE       NOMINATED NODE   READINESS GATES
openebs-ndm-bw65r                             1/1     Running            0          18h     192.168.178.152   karen      <none>           <none>
openebs-ndm-l5w8s                             1/1     Running            0          18h     192.168.178.151   josje      <none>           <none>
openebs-admission-server-698f978d67-2cwls     1/1     Running            2          18h     10.42.3.76        josje      <none>           <none>
maya-apiserver-6555665467-j6mh9               1/1     Running            0          18h     10.42.3.77        josje      <none>           <none>
cstor-disk-pool-una0-797496786c-xrtvk         3/3     Running            0          4h54m   10.42.2.3         karen      <none>           <none>
cstor-disk-pool-7it6-5b5447466f-5llc8         3/3     Running            0          5h      10.42.1.119       kristel    <none>           <none>
openebs-ndm-m76tc                             1/1     Running            0          18h     192.168.178.139   kristel    <none>           <none>
openebs-localpv-provisioner-78b7f98dd-dtz8z   1/1     Running            9          5h16m   10.42.3.78        josje      <none>           <none>
openebs-provisioner-595cdb7cc9-l7h5g          1/1     Running            9          5h5m    10.42.1.118       kristel    <none>           <none>
openebs-ndm-2q7xn                             1/1     Running            0          18h     192.168.178.138   kathleen   <none>           <none>
cstor-disk-pool-eiq0-6fdc6d9958-48qq2         3/3     Running            0          3h8m    10.42.0.5         kathleen   <none>           <none>
openebs-ndm-operator-694c6c48d9-tgxsn         0/1     CrashLoopBackOff   216        18h     10.42.3.75        josje      <none>           <none>
```
  * The CrashLoopBackOff is due to https://github.com/openebs/node-disk-manager/issues/439
* `kubectl get sc`:
```
NAME                            PROVISIONER                                                RECLAIMPOLICY   VOLUMEBINDINGMODE      ALLOWVOLUMEEXPANSION   AGE
managed-nfs-storage (default)   k8s.sjor.sg/nfs-provisioner                                Delete          Immediate              false                  216d
openebs-snapshot-promoter       volumesnapshot.external-storage.k8s.io/snapshot-promoter   Delete          Immediate              false                  52d
cstor-replicated                openebs.io/provisioner-iscsi                               Delete          WaitForFirstConsumer   false                  44d
```
* OS (from `/etc/os-release`): k3os v0.10.0 / Kubernetes v1.17.4+k3s1
* Kernel (from `uname -a`): 4.19.75-v8+

**Comment by sgielen** [comment]
The logs and command outputs of the issue description are from a cStor SPC upgrade -- but the same thing happened during a subsequent cStor volume upgrade:

```
$  kubectl logs -f -n openebs job/cstor-volume-2-1901110-20200614
I0614 16:14:22.341401       1 cstor_volume.go:99] Started upgrading cstorVolume{pvc-20bffc04-703d-44b0-8820-e124a9c0bc51} from 1.9.0 to 1.11.0
[...]
I0614 16:16:09.975421       1 cstor_volume.go:119] Successfully upgraded cstorVolume{pvc-20bffc04-703d-44b0-8820-e124a9c0bc51} from 1.9.0 to 1.11.0
I0614 16:16:09.975543       1 cstor_volume.go:99] Started upgrading cstorVolume{pvc-318b2f41-40db-4ee7-8564-4c8677665bbf} from 1.9.0 to 1.11.0
[...]
I0614 16:18:18.900259       1 cstor_volume.go:119] Successfully upgraded cstorVolume{pvc-318b2f41-40db-4ee7-8564-4c8677665bbf} from 1.9.0 to 1.11.0
I0614 16:18:18.900348       1 cstor_volume.go:99] Started upgrading cstorVolume{pvc-614627ce-bcf9-4d9e-941a-9408769679d7} from 1.9.0 to 1.11.0
I0614 16:18:20.023941       1 cstor_volume_upgrade.go:213] cstorvolumereplica pvc-614627ce-bcf9-4d9e-941a-9408769679d7-cstor-disk-pool-una0 patched
I0614 16:18:20.157426       1 cstor_volume_upgrade.go:502] Verifying the reconciliation of pvc-614627ce-bcf9-4d9e-941a-9408769679d7-cstor-disk-pool-una0 { id:3976388D6A9B1558BB9E44B60157BCD2 phase:Healthy lastTransition:2020-06-14 09:44:51 +0000 UTC }
I0614 16:18:30.303081       1 cstor_volume_upgrade.go:213] cstorvolumereplica pvc-614627ce-bcf9-4d9e-941a-9408769679d7-cstor-disk-pool-7it6 patched
I0614 16:18:30.466572       1 cstor_volume_upgrade.go:502] Verifying the reconciliation of pvc-614627ce-bcf9-4d9e-941a-9408769679d7-cstor-disk-pool-7it6 { id:FFCD17695C02A155694552333C20F9AC phase:Healthy lastTransition:2020-06-14 09:28:54 +0000 UTC }
I0614 16:18:40.486238       1 cstor_volume_upgrade.go:502] Verifying the reconciliation of pvc-614627ce-bcf9-4d9e-941a-9408769679d7-cstor-disk-pool-7it6 { id:FFCD17695C02A155694552333C20F9AC phase:Healthy lastTransition:2020-06-14 09:28:54 +0000 UTC }
I0614 16:18:50.569711       1 cstor_volume_upgrade.go:502] Verifying the reconciliation of pvc-614627ce-bcf9-4d9e-941a-9408769679d7-cstor-disk-pool-7it6 { id:FFCD17695C02A155694552333C20F9AC phase:Healthy lastTransition:2020-06-14 09:28:54 +0000 UTC }
I0614 16:19:00.650873       1 cstor_volume_upgrade.go:213] cstorvolumereplica pvc-614627ce-bcf9-4d9e-941a-9408769679d7-cstor-disk-pool-eiq0 patched
I0614 16:19:00.671937       1 cstor_volume_upgrade.go:502] Verifying the reconciliation of pvc-614627ce-bcf9-4d9e-941a-9408769679d7-cstor-disk-pool-eiq0 { id:75F84D7D04F1E97F9ECE004636C8ACB3 phase:Healthy lastTransition:2020-06-14 11:27:10 +0000 UTC }
I0614 16:19:10.851270       1 cstor_volume_upgrade.go:502] Verifying the reconciliation of pvc-614627ce-bcf9-4d9e-941a-9408769679d7-cstor-disk-pool-eiq0 { id:75F84D7D04F1E97F9ECE004636C8ACB3 phase:Healthy lastTransition:2020-06-14 11:27:10 +0000 UTC }
I0614 16:19:21.098395       1 cstor_volume_upgrade.go:581] CStor volume { phase:Healthy lastTransition:2020-06-14 09:29:05 +0000 UTC } replicaStatuses: 
		{ id:1973890494640328189 mode:Healthy chckIOSeq:1054391 r/w/s:0/0/0 uptime:24653 quorum:1 }
		{ id:10648041972266519342 mode:Healthy chckIOSeq:1054391 r/w/s:0/0/0 uptime:23675 quorum:1 }
		{ id:12576208562849748218 mode:Healthy chckIOSeq:1054391 r/w/s:0/0/0 uptime:17561 quorum:1 }
I0614 16:19:21.107948       1 helper.go:69] Patching deployment pvc-614627ce-bcf9-4d9e-941a-9408769679d7-target
E0614 16:24:24.321727       1 cstor_volume.go:113] failed to patch target deployment pvc-614627ce-bcf9-4d9e-941a-9408769679d7-target: failed to rollout because replica update in-progress: 0 of 1 updated replicas are available
F0614 16:24:24.322241       1 cstor_volume.go:77] Failed to upgrade CStor Volume pvc-614627ce-bcf9-4d9e-941a-9408769679d7:
```

Eventually, the job had three restarts. I did not capture logs for all runs, but the final one was:

```
$ kubectl logs -f -n openebs job/cstor-volume-2-1901110-20200614
I0614 16:25:38.347376       1 cstor_volume.go:99] Started upgrading cstorVolume{pvc-20bffc04-703d-44b0-8820-e124a9c0bc51} from 1.9.0 to 1.11.0
[...]
I0614 16:25:39.662453       1 cstor_volume_upgrade.go:173] cstorvolume already in 1.11.0 version
I0614 16:25:39.683143       1 cstor_volume_upgrade.go:609] Upgrade Successful for cstor volume pvc-20bffc04-703d-44b0-8820-e124a9c0bc51
I0614 16:25:39.707393       1 cstor_volume.go:119] Successfully upgraded cstorVolume{pvc-20bffc04-703d-44b0-8820-e124a9c0bc51} from 1.9.0 to 1.11.0
I0614 16:25:39.707481       1 cstor_volume.go:99] Started upgrading cstorVolume{pvc-318b2f41-40db-4ee7-8564-4c8677665bbf} from 1.9.0 to 1.11.0
[...]
I0614 16:25:41.353647       1 cstor_volume_upgrade.go:173] cstorvolume already in 1.11.0 version
I0614 16:25:41.719279       1 cstor_volume_upgrade.go:609] Upgrade Successful for cstor volume pvc-318b2f41-40db-4ee7-8564-4c8677665bbf
I0614 16:25:41.752367       1 cstor_volume.go:119] Successfully upgraded cstorVolume{pvc-318b2f41-40db-4ee7-8564-4c8677665bbf} from 1.9.0 to 1.11.0
I0614 16:25:41.752451       1 cstor_volume.go:99] Started upgrading cstorVolume{pvc-614627ce-bcf9-4d9e-941a-9408769679d7} from 1.9.0 to 1.11.0
[...]
I0614 16:25:53.643147       1 cstor_volume_upgrade.go:581] CStor volume { phase:Init lastTransition:2020-06-14 16:25:23 +0000 UTC } replicaStatuses: 
		{ id:1973890494640328189 mode:Healthy chckIOSeq:1054391 r/w/s:0/0/0 uptime:24743 quorum:1 }
		{ id:10648041972266519342 mode:Healthy chckIOSeq:1054391 r/w/s:0/0/0 uptime:23765 quorum:1 }
		{ id:12576208562849748218 mode:Healthy chckIOSeq:1054391 r/w/s:0/0/0 uptime:17651 quorum:1 }
I0614 16:25:53.775557       1 cstor_volume_upgrade.go:609] Upgrade Successful for cstor volume pvc-614627ce-bcf9-4d9e-941a-9408769679d7
I0614 16:25:53.863525       1 cstor_volume.go:119] Successfully upgraded cstorVolume{pvc-614627ce-bcf9-4d9e-941a-9408769679d7} from 1.9.0 to 1.11.0
I0614 16:25:53.863621       1 cstor_volume.go:99] Started upgrading cstorVolume{pvc-f39179d6-2121-439d-ae18-bd4bd8f47f8d} from 1.9.0 to 1.11.0
[...]
I0614 16:27:57.764532       1 cstor_volume_upgrade.go:609] Upgrade Successful for cstor volume pvc-f39179d6-2121-439d-ae18-bd4bd8f47f8d
I0614 16:27:57.864773       1 cstor_volume.go:119] Successfully upgraded cstorVolume{pvc-f39179d6-2121-439d-ae18-bd4bd8f47f8d} from 1.9.0 to 1.11.0
```

**Comment by kmova** [comment]
@sgielen could you reopen this issue if you notice this issue again. The fix in 1.12 should have addressed this.
